### PR TITLE
Fast api conversion

### LINKED ISSUE #522: Feature Request — Update API From Flask to FastAPI
**Job**
Convert API over from Flask to FastAPI and update test cases to use the FastAPI TestClient

**Describe a possible solution**
Convert to FastAPI:
- [x] '/api/class': [GET] and '/api/department': [GET] (Calvin)
- [x] '/api/subsemester': [GET] and '/api/semester': [GET] (Phoenix)
- [x] '/api/semesterInfo': [GET] and '/api/defaultsemester': [GET] (Navya)
- [x] '/api/defaultsemesterset': [POST] and '/api/mapDateRangeToSemesterPart': [POST] (Khanak)
- [x] '/api/user/<session_id>': [GET] and '/api/user': [PUT] (Changdi)
- [x] '/api/user': [POST, DELETE] (Sijin)
- [x] '/api/user/course': [POST] and '/api/event': [POST] (Yueting)
- [x] '/api/session': [POST, DELETE] and '/api/bulkCourseUpload': [POST] (Liam)
- [x] '/api/user/course': [DELETE, GET] (Sihan)

Final Touches:
- [x] Convert tests from requests to FastAPI TestClient
- [x] Remove project documents (How to Test, Project Proposal)

### PR DESCRIPTION
**Pull request by lhain08**
**Issue**

Converted API from Flask to FastAPI and wrote test cases for all endpoints

> closes #522 

**Database Changes/Migrations**

No database changes

**Test Modifications**

Added unit tests for all api endpoints, new CI workflow step for API Tests.
Tests located in src/api/tests

**Test Procedure**

Checkout this branch and run

> docker-compose -f src/api/tests/docker-compose.yml up --build --abort-on-container-exit

This will execute all tests.
You can also run the development docker containers and go to localhost:5000/docs or localhost:5000/redoc to see all documentation for the new API

**Photos**

Screen Shot 2022-04-02 at 1 20 22 AM

### CODE REVIEW

File: src/api/app.py [hunk 1]
```diff
@@ -24,139 +24,146 @@
...
     Get list of departments i.e. COGS, CIVL, CSCI, BIOL
     (Used in dropdown in "Course Search"
     """
-    semester = request.args.get("semester", default=None)
-    if semester:
-        subsemesters, error = class_info.get_subsemesters(semester)
-        return jsonify(subsemesters) if not error else Response(error, status=500)
+    if subsemester.semester:
+        subsemesters, error = class_info.get_subsemesters(subsemester.semester)
+        # for i in subsemesters:
```
Comment (jaw12346) [inline src/api/app.py:108]: Unused code should be removed.

File: src/api/app.py [hunk 2]
```diff
@@ -166,120 +173,109 @@ def uploadHandler():
...
+         if (start_dates and end_dates and semester_part_names and is_publicly_visible is not None and semester_title):
+             _, error = date_range_map.insert_all(start_dates, end_dates, semester_part_names)
+             semester_info.upsert(semester_title, is_publicly_visible)
+             if (not error):
+                 return Response(status_code=200)
+             else:
+                 return Response(error, status_code=500)
+     return Response("Did not receive proper form data", status_code=500)
+
+#@app.route('/api/user/course', methods=['GET'])
```
Comment (jaw12346) [inline src/api/app.py:205]: Unused code should be removed.

File: src/api/app.py [hunk 3]
```diff
@@ -166,120 +173,109 @@ def uploadHandler():
...
-
-    courses, error = course_select.get_selection(session['user']['user_id'])
-    return jsonify(courses) if not error else Response(error, status=500)
-
-if __name__ == '__main__':
-    app.run(debug=os.environ.get('DEBUG', 'True'), host='0.0.0.0', port=5000)
+@app.post('/api/event')
+def add_user_event(request: Request, credentials: SessionPydantic):
+    return Response(status_code=501)
+    #return event_controller.add_event(json.loads(request.data))
```
Comment (jaw12346) [inline src/api/app.py:265]: Unused code should be removed.

File: src/api/app.py [hunk 4]
```diff
@@ -166,120 +173,109 @@ def uploadHandler():
...
+    resp, error = course_select.add_selection(credentials.name, credentials.semester, request.session['user']['user_id'], credentials.cid)
+    return Response(status_code=200) if not error else Response(error, status_code=500)
+
+
+@app.delete('/api/user/course')
+async def remove_student_course(request: Request, courseDelete:CourseDeletePydantic):
+    if 'user' not in request.session:
+        return Response("Not authorized", status_code=403)
+    resp,error = course_select.remove_selection(courseDelete.name, courseDelete.semester, request.session['user']['user_id'], courseDelete.cid)
+    # resp, error = course_select.remove_selection(info['name'], info['semester'], session['user']['user_id'], info['cid'])
```
Comment (jaw12346) [inline src/api/app.py:280]: Unused code should be removed.

**Review by DarknessUnder — APPROVED** [review]

File: src/api/tests/test.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+export DB_HOST=${DB_HOST:-localhost}
+export DB_USER=${DB_USER:-yacs}
+export DB_PASS=${DB_PASS:-easy_dev_pass}
+export DB_NAME=${DB_NAME:-yacs}
+export DB_PORT=${DB_PORT:-5432}
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+export TEST_APP_DIR=$SCRIPT_DIR/../
+
+#pip3 install -r $TEST_APP_DIR/requirements.txt
```
Comment (jaw12346) [inline src/api/tests/test.sh:10]:
```suggestion
#pip3 install -r $TEST_APP_DIR/requirements.txt
```
Unused code should be removed.

File: src/api/app.py (hunk 4 above)
Comment (jaw12346) [inline src/api/app.py:280]:
```suggestion
    # resp, error = course_select.remove_selection(info['name'], info['semester'], session['user']['user_id'], info['cid'])
```
Unused code should be removed.

File: src/api/app.py (hunk 3 above)
Comment (jaw12346) [inline src/api/app.py:265]:
```suggestion
    #return event_controller.add_event(json.loads(request.data))
```
Unused code should be removed.

File: src/api/app.py (hunk 2 above)
Comment (jaw12346) [inline src/api/app.py:205]:
```suggestion
#@app.route('/api/user/course', methods=['GET'])
```
Unused code should be removed.

File: src/api/app.py (hunk 1 above)
Comment (jaw12346) [inline src/api/app.py:108]:
```suggestion
        # for i in subsemesters:
```
Unused code should be removed.

File: src/api/app.py [hunk 6]
```diff
@@ -24,139 +24,146 @@
...
     (Used in dropdown in "Course Search"
     """
-    semester = request.args.get("semester", default=None)
-    if semester:
-        subsemesters, error = class_info.get_subsemesters(semester)
-        return jsonify(subsemesters) if not error else Response(error, status=500)
+    if subsemester.semester:
+        subsemesters, error = class_info.get_subsemesters(subsemester.semester)
+        # for i in subsemesters:
+        #     print(i)
```
Comment (jaw12346) [inline src/api/app.py:109]:
```suggestion
        #     print(i)
```
Unused code should be removed.

**Review by jaw12346 — CHANGES_REQUESTED** [review]
Please check the comments that I made regarding dead code. If this remaining dead code is intentionally still in the files, you have my approval.

**Review by jaw12346 — APPROVED** [review]

**Pull request merged**